File: example-android-project/app/src/main/java/io/github/gofaith/faithdroid/FViews/FListView.java
package io.github.gofaith.faithdroid.FViews;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import org.json.JSONObject;
import org.json.JSONTokener;

import faithdroid.Faithdroid;
import io.github.gofaith.faithdroid.UI.AttrGettable;
import io.github.gofaith.faithdroid.UI.AttrSettable;
import io.github.gofaith.faithdroid.UI.UIController;

public class FListView extends FView implements AttrGettable,AttrSettable{
    private RecyclerView v;
    private MyAdapter adapter;
    public String fnOnGetView,fnOnBindData,fnOnGetCount;
    public String onItemClick,onItemLongClick;
    public FListView(UIController uiController, RecyclerView.LayoutManager lm) {
        parentController =uiController;
        v = new RecyclerView(parentController.activity);
        view=v;
        v.setLayoutManager(lm);
        adapter=new MyAdapter(this);
        v.setAdapter(adapter);
    }

    @Override
    public String getAttr(String attr) {
        String str = getUniversalAttr(attr);
        if (str != null) {
            return str;
        }
        switch (attr){
        }
        return null;
    }

    @Override
    public void setAttr(String attr, final String value) {
        if (setUniversalAttr(attr, value)) {
            return;
        }
        switch (attr){
            // --------------------------------------------
            case "OnGetView":
                fnOnGetView=value;
                break;
            case "OnBindData":
                fnOnBindData=value;
                break;
            case "OnGetCount":
                fnOnGetCount = value;
                break;
            case "NotifyDataSetChanged":
                v.getAdapter().notifyDataSetChanged();
                break;
            case "OnItemClick":
                onItemClick=value;
                break;
            case "OnItemLongClick":
                onItemLongClick=value;
                break;
        }
        return ;
    }
    class MyAdapter extends RecyclerView.Adapter<MyAdapter.ViewHolder> {
        private FListView parentListView;
        private OnItemClickListener mItemClickListener=new OnItemClickListener() {
            @Override
            public void onItemClick(View view) {
                if (parentListView.onItemClick==null||parentListView.onItemClick.equals(""))
                    return;
                int pos=parentListView.v.getChildAdapterPosition(view);
                Faithdroid.triggerEventHandler(parentListView.onItemClick,String.valueOf(pos));
            }

            @Override
            public void onItemLongClick(View view) {
                if (parentListView.onItemLongClick==null||parentListView.onItemLongClick.equals(""))
                    return;
                int pos=parentListView.v.getChildAdapterPosition(view);
                Faithdroid.triggerEventHandler(parentListView.onItemLongClick,String.valueOf(pos));
            }
        };

        public class ViewHolder extends RecyclerView.ViewHolder {
            public FView fView;
            public String str;
            public ViewHolder(FView v,String str) {
                super(v.view);
                fView = v;
                this.str =str;
            }
        }
        public MyAdapter(FListView fListView) {
            parentListView =fListView;
        }
        @Override
        public MyAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            if (parentListView.fnOnGetView == null) {
                return new ViewHolder(null,"");
            }
            String viewGot= Faithdroid.triggerEventHandler(parentListView.fnOnGetView,"");

            try {
                JSONObject object = (JSONObject) (new JSONTokener(viewGot).nextValue());
                FView v = parentListView.parentController.viewmap.get(object.getString("VID"));
                v.view.setLayoutParams(FFrameLayout.parseLP(v));
                MyAdapter.ViewHolder vh = new ViewHolder(v, viewGot);
                vh.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mItemClickListener.onItemClick(v);
                    }
                });
                vh.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        mItemClickListener.onItemLongClick(v);
                        return true;
                    }
                });
                return vh;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }
        @Override
        public void onBindViewHolder(ViewHolder holder, int position) {
            if (parentListView.fnOnBindData == null) {
                return;
            }
            try {
                JSONObject jo = new JSONObject();
                jo.put("Str", holder.str);
                jo.put("Position", position);
                Faithdroid.triggerEventHandler(parentListView.fnOnBindData, jo.toString());
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        @Override
        public int getItemCount() {
            if (parentListView.fnOnGetCount == null) {
                return 0;
            }
            try {
                return Integer.parseInt(Faithdroid.triggerEventHandler(parentListView.fnOnGetCount,""));
            } catch (Exception e) {
                e.printStackTrace();
            }
            return 0;
        }
    }
    public static interface OnItemClickListener {
        void onItemClick(View view);
        void onItemLongClick(View view);
    }
}

File: android/app/src/main/java/io/github/stevenzack/faithdroidapp/FViews/FListView.kt
package io.github.stevenzack.faithdroidapp.FViews

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import faithdroid.Faithdroid
import io.github.stevenzack.faithdroidapp.UI.AttrGettable
import io.github.stevenzack.faithdroidapp.UI.AttrSettable
import io.github.stevenzack.faithdroidapp.UI.FView
import io.github.stevenzack.faithdroidapp.UI.UIController
import org.json.JSONObject
import org.json.JSONTokener


class FListView(uiController: UIController, lm: RecyclerView.LayoutManager) : FView(), AttrGettable, AttrSettable {
    private val v: RecyclerView
    private val adapter: MyAdapter
    var fnOnGetView: String? = null
    var fnOnBindData: String? = null
    var fnOnGetCount: String? = null
    var onItemClick: String? = null
    var onItemLongClick: String? = null

    init {
        parentController = uiController
        v = RecyclerView(parentController!!.activity)
        view = v
        v.setLayoutManager(lm)
        adapter = MyAdapter(this)
        v.setAdapter(adapter)
    }

    override  fun getAttr(attr: String): String {
        val str = getUniversalAttr(attr)
        if (str != null) {
            return str
        }
        when (attr) {

        }
        return ""
    }

    override fun setAttr(attr: String, value: String) {
        if (setUniversalAttr(attr, value)) {
            return
        }
        when (attr) {
            // --------------------------------------------
            "OnGetView" -> fnOnGetView = value
            "OnBindData" -> fnOnBindData = value
            "OnGetCount" -> fnOnGetCount = value
            "NotifyDataSetChanged" -> v.getAdapter()!!.notifyDataSetChanged()
            "OnItemClick" -> onItemClick = value
            "OnItemLongClick" -> onItemLongClick = value
        }
        return
    }

    inner class MyAdapter(private val parentListView: FListView) :
        RecyclerView.Adapter<MyAdapter.ViewHolder>() {
        override fun getItemCount(): Int {
            if (parentListView.fnOnGetCount == null) {
                return 0
            }
            try {
                return Integer.parseInt(Faithdroid.triggerEventHandler(parentListView.fnOnGetCount, ""))
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return 0
        }

        private val mItemClickListener = object : OnItemClickListener {
            override fun onItemClick(view: View) {
                if (parentListView.onItemClick == null || parentListView.onItemClick == "")
                    return
                val pos = parentListView.v.getChildAdapterPosition(view)
                Faithdroid.triggerEventHandler(parentListView.onItemClick, pos.toString())
            }

            override fun onItemLongClick(view: View) {
                if (parentListView.onItemLongClick == null || parentListView.onItemLongClick == "")
                    return
                val pos = parentListView.v.getChildAdapterPosition(view)
                Faithdroid.triggerEventHandler(parentListView.onItemLongClick, pos.toString())
            }
        }

        inner class ViewHolder(var fView: FView, var str: String) : RecyclerView.ViewHolder(fView.view!!)

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            if (parentListView.fnOnGetView == null) {
                return ViewHolder(null!!, "")
            }
            val viewGot = Faithdroid.triggerEventHandler(parentListView.fnOnGetView, "")
            val `object` = JSONTokener(viewGot).nextValue() as JSONObject
            val v = parentListView.parentController!!.viewmap.get(`object`.getString("VID"))
            v!!.view!!.setLayoutParams(FFrameLayout.parseLP(v))
            val vh = ViewHolder(v, viewGot)
            vh.itemView.setOnClickListener(object : View.OnClickListener{
                override fun onClick(v: View) {
                    mItemClickListener.onItemClick(v)
                }
            })
            vh.itemView.setOnLongClickListener(object : View.OnLongClickListener {
                override fun onLongClick(v: View): Boolean {
                    mItemClickListener.onItemLongClick(v)
                    return true
                }
            })
            return vh
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            if (parentListView.fnOnBindData == null) {
                return
            }
            try {
                val jo = JSONObject()
                jo.put("Str", holder.str)
                jo.put("Position", position)
                Faithdroid.triggerEventHandler(parentListView.fnOnBindData, jo.toString())
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }

    interface OnItemClickListener {
        fun onItemClick(view: View)
        fun onItemLongClick(view: View)
    }
}
